File: BEmu/MarketDataRequest/ElementMarketSubscriptionCancelReason.cs
//------------------------------------------------------------------------------
// <copyright project="BEmu" file="MarketDataRequest/ElementMarketSubscriptionCancelReason.cs" company="Jordan Robinson">
//     Copyright (c) 2013 Jordan Robinson. All rights reserved.
//
//     The use of this software is governed by the Microsoft Public License
//     which is included with this distribution.
// </copyright>
//------------------------------------------------------------------------------

namespace BEmu.MarketDataRequest
{
    using System;
    using System.Collections.Generic;
    using System.Linq;
    using System.Text;

    internal class ElementMarketSubscriptionCancelReason : Element
    {
        private readonly ElementMarketString _source, _category, _description, _subCategory;
        private readonly ElementMarketInt _errorCode;

        internal ElementMarketSubscriptionCancelReason()
        {
            this._source = new ElementMarketString("source", "Session");
            this._errorCode = new ElementMarketInt("errorCode", 0);
            this._category = new ElementMarketString("category", "CANCELED");
            this._description = new ElementMarketString("description", "Subscription canceled");
            this._subCategory = new ElementMarketString("subcategory", "");
        }

        public override Name Name { get { return new Name("reason"); } }
        public override int NumElements { get { return 5; } }
        public override int NumValues { get { return 1; } }
        public override bool IsArray { get { return false; } }
        public override bool IsComplexType { get { return true; } }
        public override bool IsNull { get { return false; } }

        public override object this[int index] { get { return null; } }

        public override Element this[string name]
        {
            get
            {
                foreach (var item in this.Elements)
                {
                    if (item.Name.ToString().ToUpper() == name.ToUpper())
                        return item;
                }
                return base[name];
            }
        }

        public override IEnumerable<Element> Elements
        {
            get
            {
                yield return this._source;
                yield return this._errorCode;
                yield return this._category;
                yield return this._description;
                yield return this._subCategory;
            }
        }

        internal override StringBuilder PrettyPrint(int tabIndent)
        {
            string tabs = Types.IndentType.Indent(tabIndent);
            StringBuilder result = new StringBuilder();

            result.AppendFormat("{0}reason = {{{1}", tabs, Environment.NewLine);
            foreach (var item in this.Elements)
            {
                result.Append(item.PrettyPrint(tabIndent + 1));
            }
            result.AppendFormat("{0}}}{1}", tabs, Environment.NewLine);
            return result;
        }
    }
}

File: BEmuJava/bemu/src/main/java/com/bemu/BEmu/MarketDataRequest/ElementMarketSubscriptionCancelReason.java
//------------------------------------------------------------------------------
// <copyright project="BEmuJava" file="BEmu.MarketDataRequest.ElementMarketSubscriptionCancelReason.java" company="Jordan Robinson">
//     Copyright (c) 2013 Jordan Robinson. All rights reserved.
//
//     The use of this software is governed by the Microsoft Public License
//     which is included with this distribution.
// </copyright>
//------------------------------------------------------------------------------

package com.bemu.BEmu.MarketDataRequest;

import com.bemu.BEmu.Name;
import com.bemu.BEmu.Element;

import java.util.List;
import java.util.LinkedList;

public class ElementMarketSubscriptionCancelReason extends ElementParent
{
    private final ElementMarketString _source, _category, _description, _subCategory;
    private final ElementMarketInt _errorCode;

    ElementMarketSubscriptionCancelReason()
    {
        this._source = new ElementMarketString("source", "Session");
        this._errorCode = new ElementMarketInt("errorCode", 0);
        this._category = new ElementMarketString("category", "CANCELED");
        this._description = new ElementMarketString("description", "Subscription canceled");
        this._subCategory = new ElementMarketString("subcategory", "");
    }

    public Name name()
    {
    	return new Name("reason");
    }
    
    public int numValues()
    {
    	return 1;
    }
    
    public int numElements()
    {
    	return 5;
    }
    
    public boolean isComplexType()
    {
    	return true;
    }
    
    public boolean isArray()
    {
    	return false;
    }
    
    public boolean isNull()
    {
    	return false;
    }
    
    private List<Element> Elements()
    {
    	List<Element> result = new LinkedList<Element>();
    	result.add(this._source);
    	result.add(this._errorCode);
    	result.add(this._category);
    	result.add(this._description);
    	result.add(this._subCategory);
    	return result;
    }
    
    protected StringBuilder prettyPrint(int tabIndent)
    {
        String tabs = com.bemu.BEmu.types.IndentType.Indent(tabIndent);
        StringBuilder result = new StringBuilder();

        List<Element> elements = this.Elements(); 
        result.append(String.format("%sreason = {%s", tabs, System.getProperty("line.separator")));
        for(int i = 0; i < elements.size(); i++)
        {
        	ElementParent elm = (ElementParent)elements.get(i);
        	result.append(elm.prettyPrint(tabIndent + 1));
        }
        result.append(String.format("%s}%s", tabs, System.getProperty("line.separator")));

        return result;
    }
    
}
